klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+                    rule_counts_df["rule_name"] = rule_counts_df["analysis_rule"].str.split(RULE_STR_SEP).str[0]
+                    rule_counts_df["type"] = rule_counts_df["analysis_rule"].str.split(RULE_STR_SEP).str[1]
+                    rule_counts_df.drop(columns=["analysis_rule"], inplace=True)
+                    rule_counts_df["count"] = rule_counts_df["count"].astype(int)
+                    rule_counts_df = rule_counts_df.pivot(index="rule_name",
+                                                          columns="type",
+                                                          values="count")
+                    rule_counts_df = rule_counts_df.fillna(0)
+                    rule_counts_df = rule_counts_df.reset_index(drop=False)
+                    rule_counts_df = rule_counts_df.rename(columns={"index": "rule_name"})
+
+                    rule_counts_df["table_name"] = tc_df.loc[tc_df["run_id"] == run_id, "table_name"].iloc[0]
+
+                    for c in ANALYSIS_RULES:
+                        if c not in rule_counts_df.columns:
+                            rule_counts_df[c] = 0
+
+                    rule_counts_df[ANALYSIS_RULES] = rule_counts_df[ANALYSIS_RULES].astype(int)
+                    rule_counts_df = rule_counts_df[["table_name", "rule_name"] + ANALYSIS_RULES]
+                    results_df_list.append(rule_counts_df)
+                    tc_df.loc[tc_df["run_id"] == run_id, "run_status"] = "completed"
+                else:
+                    tc_df.loc[tc_df["run_id"] == run_id, "run_status"] = "no output"
+
+            except Exception as exc:
+                logging.error("%d generated an exception: %s" % (run_id, exc))
+                traceback.print_exc()
+                tc_df.loc[tc_df["run_id"] == run_id, "run_status"] = "exception"
+
+    if len(results_df_list) > 0:
+        results_df = pd.concat(results_df_list)
+    else:
+        results_df = pd.DataFrame()
+
+    return results_df, tc_df
+
+
+def parse_existing_rules(rule_deck_path, output_path, target_table=None):
+    """
+    This function collects the rule names from the existing drc rule decks.
+
+    Parameters
+    ----------
+    rule_deck_path : string or Path object
+        Path string to the DRC directory where all the DRC files are located.
+    output_path : string or Path
+        Path of the run location to store the output analysis file.
+    target_table : string Optional
+        Name of the table to be in testing
+
+    Returns
+    -------
+    pd.DataFrame
+        A pandas DataFrame with the rule and rule deck used.
+    """
+
+    if target_table is None:
+        drc_files = glob.glob(os.path.join(rule_deck_path, "rule_decks", "*.drc"))
+    else:
+        table_rule_file = os.path.join(rule_deck_path, "rule_decks", f"{target_table}.drc")
+        if not os.path.isfile(table_rule_file):
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), table_rule_file)
```

Comment:
+1 for surfacing structured exception here.

Reply:
Trying to follow your lead and taking in your requirements.